README.md
# Virtual Assistant

Welcome to the Virtual Assistant project repository
![Virtual Assistant](./static/hello.gif)  

#### **SIGNIFICANT INFO**
**I will not answer any of your questions if you did not read the Wiki**  
You can find all information about the system in the [Wiki Page](https://github.com/BiggyBaron/VirtualAssistant/wiki)   
**Seriously, read it, please. I spend a week to write it**

## Passport of the project

Virtual Assistant is the project of the Erasmus-Plus program students in the Middlesex University, London.
The main idea of the system is to control Smart-Home environment with dialogue based control interface.
The system is developed to to assists people with different preferences in their Activities of Daily Life (ADL).
The system uses Graphical User Interface with text based dialogue
window; speech recognition and text to speech generating software (Google API). 
Virtual Assistant uses pre-defined preferences based on age of the user profile
with differentiated priorities of the needs including health, food, work, entertainment, security and
energy consumption of the devices. Virtual Assistant passed the validation tests done by
the developers and showed acceptable results in conflict resolution. The system uses Case-Based
Reasoning module to resolve conflicts, Feed-Forward Artificial Neural Networks to classify 
orders of the users and the reasons of the orders. Place of the research Sunny Hill House, Middlesex University, London.

![VirtAss demo page](./images/mockup.png)  
Interface of the project.

## Getting Started

These instructions will get you a copy of the project up and running on your local machine for development and testing purposes. See deployment for notes on how to deploy the project on a live system.

### Prerequisites

1. Linux Uubuntu 16.04 (can be used any other Linux, but with other steps below)
2. Git 2.7.4.
```Shell
sudo apt install git
```
3. Python 3.6 and PIP for Python 3.6.
```Shell
sudo apt install python3-dev python3-pip
```
4. Copy repository of the system to your machine:
```Shell
git clone https://github.com/BiggyBaron/VirtualAssistant.git
cd VirtualAssistant
```
5. Install all libraries:
```Shell
sudo pip3 install -r requirements.txt
```
6. Download Natural Language Toolkit database:
```Shell
python3 install_nltk.py
```

### Installing MySQL
1. Update package index and packages:
```Shell
sudo apt-get update
sudo apt-get upgrade
```
2. Install MySQL:
```Shell
sudo apt-get install mysql-server
```
3. Set up MySQL (user=root, password="123"):
```Shell
sudo mysql_secure_installation
```
4. Check if MySQL is running:
```Shell
systemctl status mysql.service
```
5. Create MySQL database "virtass":
* Enter to MySQL command line (password=123):
```Shell
mysql -u root -p
```
* Enter MySQL command to create it:
```SQL
CREATE DATABASE virtass;
```
6. Import database from repository:
```Shell
cd VirtualAssistant/DataBaseBackup
mysql -u root -p virtass < virtass.sql
```

## Running all system

To run all system, go to the repository folder and start main.py:
```Shell
cd VirtualAssistant
python3 main.py
```
In browser open:  
https://localhost:8090/  
Or:  
https://your_ip:8090/  

## Built With

* [Flask web framework](https://www.jetbrains.com/pycharm/)
* [Lucid chart diagram online editor](https://www.lucidchart.com)
* [Pycharm IDE](https://www.jetbrains.com/pycharm/)
* [Una animated emojies](https://github.com/una/animated-emojis)
* [W3School CSS templates](https://www.w3schools.com/w3css/w3css_templates.asp)
* [Smart Home Controller VeraSecure](http://getvera.com/)

## Authors

* **Bauyrzhan Ospan**
* **Mario Jose Quinde Li Say Tan**
* **Kenzhegali Nurgaliyev**

## License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details

## Acknowledgments

* Diseases and condition list is from https://www.cdc.gov/diseasesconditions/index.html
* Perceptron made on work of http://www.tarnlab.com/text-classification-using-neural-networks/

## Bibliography

* [Improved Multi-user Interaction in
a Smart Environment Through a Preference-Based Conflict Resolution Virtual Assistant](http://ieeexplore.ieee.org/document/8114654/)
* [A Sensitivity Analysis of (and Practitioners' Guide to) Convolutional Neural Networks for Sentence Classification](https://arxiv.org/abs/1510.03820)
* [Using argumentation to manage users’ preferences](https://www.sciencedirect.com/science/article/pii/S0167739X1630379X)
* [Is Context-aware Reasoning = Case-based Reasoning?](http://eprints.mdx.ac.uk/21527/)
* [A survey on managing users’ preferences in ambient intelligence](https://link.springer.com/article/10.1007/s10209-017-0527-y)
* [Learning frequent behaviours of the users in 
Intelligent Environments](http://ieeexplore.ieee.org/document/6516530/)
* [Distance Metric Learning for Large Margin Nearest Neighbor Classification](http://www.jmlr.org/papers/v10/weinberger09a.html)
* [Convolutional Neural Networks for Sentence Classification](https://arxiv.org/abs/1408.5882)
* [An Agent-Based Architecture for Sensor Data Collection and Reasoning in Smart Home Environments for Independent Living](https://dl.acm.org/citation.cfm?id=2966058)
* [Design and evaluation of a smart home voice interface for the elderly: acceptability and objection aspects](https://link.springer.com/article/10.1007/s00779-011-0470-5)
* [Configuration of smart environments made simple: Combining visual modeling with semantic metadata and reasoning](http://ieeexplore.ieee.org/document/7030116/)
* [Towards adaptive control in smart homes: Overall system design and initial evaluation of activity recognition](https://dspace.lboro.ac.uk/dspace-jspui/handle/2134/24473)
* [Machine Learning Based Adaptive Context-Aware System for Smart Home Environment](https://pdfs.semanticscholar.org/8cf5/fe5062727744f5429bb34d9c0bd24f439ee6.pdf)
* [An Ontology-based Context-aware System for Smart Homes: E-care@home](https://www.ncbi.nlm.nih.gov/pubmed/28684686)
* [Policy Conflict Resolution in IoT via Planning](https://link.springer.com/chapter/10.1007/978-3-319-57351-9_22)
* [Context-based conflict management in pervasive platforms](http://ieeexplore.ieee.org/document/7917567/)
* [SVM-Based Multimodal Classification of Activities of Daily Living in Health Smart Homes: Sensors, Algorithms, and First Experimental Results](http://ieeexplore.ieee.org/abstract/document/5352277/)
* [Context-Aware User Modeling and Semantic Interoperability in Smart Home Environments](http://ieeexplore.ieee.org/document/6735563/)
* [Text Categorization by Backpropagation Network](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.206.4349&rep=rep1&type=pdf)
* [Simulation of a Simple Bio-Mimetic Robot with Neuromorphic Control System and Optimization Based on the Genetic Algorithm](http://ijiet.com/wp-content/uploads/2017/09/9.pdf)

Report/history.txt
Version history for the LLNCS LaTeX2e class

 date     filename      version   action/reason/acknowledgements
----------------------------------------------------------------------------
 29.5.96  letter.txt      beta    naming problems (subject index file)
                                  thanks to Dr. Martin Held, Salzburg, AT

          subjindx.ind            renamed to subjidx.ind as required
                                  by llncs.dem

          history.txt             introducing this file

 30.5.96  llncs.cls               incompatibility with new article.cls of
                                  1995/12/20 v1.3q Standard LaTeX document class,
                                  \if@openbib is no longer defined,
                                  reported by Ralf Heckmann and Graham Gough
                                  solution by David Carlisle

 10.6.96  llncs.cls               problems with fragile commands in \author field
                                  reported by Michael Gschwind, TU Wien

 25.7.96  llncs.cls               revision a corrects:
                                  wrong size of text area, floats not \small,
                                  some LaTeX generated texts
                                  reported by Michael Sperber, Uni Tuebingen

 16.4.97  all files        2.1    leaving beta state,
                                  raising version counter to 2.1

  8.6.97  llncs.cls        2.1a   revision a corrects:
                                  unbreakable citation lists, reported by
                                  Sergio Antoy of Portland State University

11.12.97  llncs.cls        2.2    "general" headings centered; two new elements
                                  for the article header: \email and \homedir;
                                  complete revision of special environments:
                                  \newtheorem replaced with \spnewtheorem,
                                  introduced the theopargself environment;
                                  two column parts made with multicol package;
                                  add ons to work with the hyperref package

07.01.98  llncs.cls        2.2    changed \email to simply switch to \tt

25.03.98  llncs.cls        2.3    new class option "oribibl" to suppress
                                  changes to the thebibliograpy environment
                                  and retain pure LaTeX codes - useful
                                  for most BibTeX applications

16.04.98  llncs.cls        2.3    if option "oribibl" is given, extend the
                                  thebibliograpy hook with "\small", suggested
                                  by Clemens Ballarin, University of Cambridge

20.11.98  llncs.cls        2.4    pagestyle "titlepage" - useful for
                                  compilation of whole LNCS volumes

12.01.99  llncs.cls        2.5    counters of orthogonal numbered special
                                  environments are reset each new contribution

27.04.99  llncs.cls        2.6    new command \thisbottomragged for the
                                  actual page; indention of the footnote
                                  made variable with \fnindent (default 1em);
                                  new command \url that copys its argument

 2.03.00  llncs.cls        2.7    \figurename and \tablename made compatible
                                  to babel, suggested by Jo Hereth, TU Darmstadt;
                                  definition of \url moved \AtBeginDocument
                                  (allows for url package of Donald Arseneau),
                                  suggested by Manfred Hauswirth, TU of Vienna;
                                  \large for part entries in the TOC

16.04.00  llncs.cls        2.8    new option "orivec" to preserve the original
                                  vector definition, read "arrow" accent

17.01.01  llncs.cls        2.9    hardwired texts made polyglot,
                                  available languages: english (default),
                                  french, german - all are "babel-proof"

20.06.01  splncs.bst              public release of a BibTeX style for LNCS,
                                  nobly provided by Jason Noble

14.08.01  llncs.cls        2.10   TOC: authors flushleft,
                                  entries without hyphenation; suggested
                                  by Wiro Niessen, Imaging Center - Utrecht

23.01.02  llncs.cls        2.11   fixed footnote number confusion with
                                  \thanks, numbered institutes, and normal
                                  footnote entries; error reported by
                                  Saverio Cittadini, Istituto Tecnico
                                  Industriale "Tito Sarrocchi" - Siena

28.01.02  llncs.cls        2.12   fixed footnote fix ; error reported by
                                  Chris Mesterharm, CS Dept. Rutgers - NJ

28.01.02  llncs.cls        2.13   fixed the fix (programmer needs vacation)

17.08.04  llncs.cls        2.14   TOC: authors indented, smart \and handling
                                  for the TOC suggested by Thomas Gabel
                                  University of Osnabrueck

07.03.06  splncs.bst              fix for BibTeX entries without year; patch
                                  provided by Jerry James, Utah State University

14.06.06  splncs_srt.bst          a sorting BibTeX style for LNCS, feature
                                  provided by Tobias Heindel, FMI Uni-Stuttgart

16.10.06  llncs.dem        2.3    removed affiliations from \tocauthor demo

11.12.07  llncs.doc               note on online visibility of given e-mail address

15.06.09  splncs03.bst            new BibTeX style compliant with the current

main.py
#!/usr/bin/python3
# -*- coding: UTF-8 -*-
import random
import urllib

import pymysql
from flask import Flask, render_template, request, Markup

import DM

app = Flask(__name__)  # Creating new flask app


def write2dbLOG(key, value):  # This is database logging engine, it takes key and value and stores it in eval table

    # Connecting virtass database
    con = pymysql.connect(host='0.0.0.0', unix_socket='/tmp/mysql.sock', user="root", passwd="123", db='virtass')
    cur = con.cursor(pymysql.cursors.DictCursor)
    # Connecting to the table eval and changing given value with given key
    id = cur.execute("SELECT * FROM `eval` ORDER BY `eval`.`id` ASC LIMIT 10000")
    cur.execute("UPDATE `eval` SET `" + str(key) + "` = '" + str(value)
                + "' WHERE `eval`.`id` = '" + str(id) + "';")
    # Commit and push changes, close database
    con.commit()
    cur.close()
    con.close()


# This module checks status of the devices
# It sends get request to Vera and parses it
def check_status(deviceNum):  # Taking device number to parse its value
    # Sending GET request to the Vera
    # IP of Vera was 10.12.102.156 - if it changed, change it here
    data = urllib.request.urlopen("http://10.12.102.156/port_3480/data_request?id=lu_status").read()

    d = data.decode("utf-8")  # Decoding request
    # Parse status of the device
    try:
        Status = int(d.split('"id": ' + str(deviceNum))[1].split('" }', 1)[0].strip()[-1])
        states = ["off", "on"]
        return states[Status]
    except:
        return "somehow?"


# This is randomiser module
# It creates scenarios for users inside the Evaluation test
def randomiser():
    # It connect to the Database
    con = pymysql.connect(host='0.0.0.0', unix_socket='/tmp/mysql.sock', user="root", passwd="123", db='virtass')
    cur = con.cursor(pymysql.cursors.DictCursor)
    # Block of software that creates random conflict
    users = ["Father", "Mother", "Son", "Grandpa"]
    usertypes2 = {"Father": "adult", "Mother": "adult", "Son": "young", "Grandpa": "elder"}
    reasons = ["energy", "entertainment", "food", "health", "security", "work"]
    devices = ["lamp"]
    u = random.SystemRandom().choice(users)  # Crypto-secure random choice
    user = usertypes2[u]
    users2 = users
    users2.remove(u)
    u2 = random.SystemRandom().choice(users2)
    user2 = usertypes2[u2]
    reason = random.SystemRandom().choice(reasons)
    dev = random.SystemRandom().choice(devices)
    # It takes GET from Vera to create scenario which is real such that device in the status of rule
    data = urllib.request.urlopen("http://10.12.102.156/port_3480/data_request?id=lu_status").read()
    d = data.decode("utf-8")
    try:
        if dev == "kettle":
            device = 19
        else:
            device = 395
        status = int(d.split('"id": ' + str(device))[1].split('" }', 1)[0].strip()[-1])
    except ValueError:
        status = 0
    # Commit changes and close database
    cur.close()
    con.close()
    # Again connection to the Database, I don`t know why
    con = pymysql.connect(host='0.0.0.0', unix_socket='/tmp/mysql.sock', user="root", passwd="123", db='virtass')
    cur = con.cursor(pymysql.cursors.DictCursor)
    # Update of information inside the rules table
    cur.execute("UPDATE `rules` SET `user` = '" + str(user2) + "' WHERE `rules`.`id` = 0")
    con.commit()
    cur.execute("UPDATE `rules` SET `reason` = '" + str(reason) + "' WHERE `rules`.`id` = 0")
    con.commit()
    cur.execute("UPDATE `rules` SET `device` = '" + str(dev) + "' WHERE `rules`.`id` = 0")
    con.commit()
    cur.execute("UPDATE `rules` SET `status` = '" + str(status) + "' WHERE `rules`.`id` = 0")
    # Commit and close database
    con.commit()
    cur.close()
    con.close()

    # Add values to eval table, updating evaluation tables by data created before by Randomizer
    con = pymysql.connect(host='0.0.0.0', unix_socket='/tmp/mysql.sock', user="root", passwd="123", db='virtass')
    cur = con.cursor(pymysql.cursors.DictCursor)
    order = dev.title() + ["On", "Off"][status]
    Newreason = random.SystemRandom().choice(reasons)
    cur.execute("INSERT INTO `eval` (`user`, `otheruser`, `ordertype`, `rulereason`, `reason`) VALUES ('" +
                str(user) + "', '" + str(user2) + "', '" + str(order) + "', '" + str(reason) + "', '" + str(
        Newreason) + "');")
    # Commit and close
    con.commit()
    id23 = cur.execute("SELECT * FROM `eval` ORDER BY `eval`.`id` ASC LIMIT 10000")
    cur.close()
    con.close()

    # Write to answer.txt new Evaluation Instruction
    priorities = {"adult": "highest", "elder": "second highest", "young": "lowest"}
    preferences = {"adult": "1: Security, 2: Health, 3: Work, 4: Food, 5: Energy, 6: Entertainment",
                   "elder": "1: Health, 2: Security, 3: Work, 4: Food, 5: Energy, 6: Entertainment",
                   "young": "1: Security, 2: Work, 3: Entertainment, 4: Health, 5: Food, 6: Energy"}
    priority = priorities[user]
    preference = preferences[user]
    priority2 = priorities[user2]
    preference2 = preferences[user2]

    answer = "Your ID is " + str(id23) + "<br>" + \
             "The system can be used by different users (Father, Mother, Son, Grandpa) that have different " \
             "roles (Adult, Elder, Young).<br>Please, go through the instructions. " \
             "<br><br><h4>Instructions:</h4>" \
             "The system has randomly chosen " \
             "to be user '"
    answer = answer + str(u) + "'.<br>" + str(u) + " is the '" + str(user) + "' usertype which is the " \
             + str(priority) + " priority usertype.<br>" + str(u) + " has the " \
                                                                    "following preferences (1 is the highest):<br>" \
             + str(preference) + "<br>"

    answer = answer + "The other user of the system is the '" + str(u2) + "' with usertype '" + str(
        user2) + "' which is the " \
             + str(priority2) + " priority usertype.<br> The other user '" + str(u2) + "' has the " \
                                                                                       "following preferences:<br>" \
             + str(preference2) + "<br><br>" + "<h5>Conflict:</h5><br>There is a conflict between you (" + \
             str(u) + ":" + str(user) + ") and the other " \
                                        "user (" + str(u2) + ":" + str(
        user2) + ") .<br>The other user turned the " + str(dev) + " " + \
             str(["off", "on"][status]) + " with a reason based on the " + str(reason) + " issue.<br>" + \
             "But you need the device to be in an opposite status because your reason is " + str(Newreason) \
             + " issue.<br> Let`s check how the system will resolve this conflict."
    # Writing to manual.txt new Evaluation test manual
    Manual = "<ol><li>On the login page choose '" + str(u) + "'. </li>" + \
             "<li>Then type or say (by clicking the microphone) in your own words the order to turn " \
             + str(["on", "off"][status]) + " the " + str(dev) + ".</li>" \
             + "<li>Then the system will inform you that there is a conflict and will ask you about " \
               "the next step. Type or say 'yes'." + "</li>" \
             + "<li>Then type or say (by clicking the microphone) in your own words the reason about the " + \
             str(Newreason) + " issue.</li>" + "<li>Then the system resolves the conflict and " \
                                               "you will need to answer the questionnaire." \
                                               "</li></ol>"
    # Creating head of the summary.txt of the evaluation test
    summary = "<h2>Story:</h2><ol><li>Your user is " + str(u) + ", " + str(user) + " usertype (" + str(priority) + \
              " priority) with next preferences:<br>" + str(preference) + "<br></li><li>" + \
              "Other user is " + str(u2) + ", " + str(user2) + " usertype (" + str(priority2) + \
              " priority) with next preferences:<br>" + str(preference2) + "<br></li><li>" + \
              "Conflict: you (" + str(u) + ":" + str(user) + ") want to turn the " + str(dev) + \
              " " + str(["on", "off"][status]) + \
              " but other user (" + str(u2) + ":" + str(user2) + ") did the opposite. </li><li>" + \
              "Other user`s (" + str(u2) + ", " + str(user2) + ") reason is " + str(reason).upper() + \
              "</li><li>Your (" + str(u) + ", " + str(user) + ") reason is: "
    # Writing them down
    with open("answer.txt", 'w') as out:
        out.write(answer)
    with open("manual.txt", 'w') as o:
        o.write(Manual)
    with open("summary.txt", 'w') as o:
        o.write(summary)


# Function to write summary of the Evaluation Test in the end of test
def Summary():
    # Writing header for flask, it will be rendered inside the page
    header = "Summary"
    # Read previously created summary and store it in variable
    with open("summary.txt", 'r') as out:
        summary = out.read()
    # Connect to eval table and take all data about evaluation test
    con = pymysql.connect(host='0.0.0.0', unix_socket='/tmp/mysql.sock', user="root", passwd="123", db='virtass')
    cur = con.cursor(pymysql.cursors.DictCursor)
    cur.execute('SELECT * FROM `eval`')
    dataset = cur.fetchall()
    # Close table
    cur.close()
    con.close()
    # Parse variables from table
    orderraw = dataset[-1]["orderraw"]
    orderdef = dataset[-1]["orderdef"]
    reasonraw = dataset[-1]["reasonraw"]
    reasondef = dataset[-1]["reasondef"]
    output = dataset[-1]["output"]
    winner = [" other user`s ", " your "][output]
    # Write end of the summary.txt and send it to flask.render
    summary = summary + str(reasondef).upper() + "</li>" + "<li>System desided that" + str(winner) + \
              "order has a higher priority.</li></ol><h2>Classification:</h2><ol><li>Your order was: " + \
              str(orderraw) + "</li><li>System classified order as: " + str(orderdef) + "</li><li>" + \
              "Your reason was: " + str(reasonraw) + "</li><li>" + \
              "System classified reason as: " + str(reasondef) + "</li></ol>"
    answer = Markup(summary)  # Markup used to make the text readable by browser
    return answer, header  # Returning header and info for summary page


# Reading answer.txt and returning it to flask.render after markdown
def readANS():
    with open("answer.txt", 'r') as out:
        answer = out.read()
    Info = Markup(answer)
    header = "Welcome to the evaluation test"
    return Info, header


# Reading manual.txt and returning it to flask.render after markdown
def readMan():
    with open("manual.txt", 'r') as out:
        answer = out.read()
    Info = Markup(answer)
    header = "Manual"
    return Info, header


# Login page, no authorisation with password
# TODO: add password
@app.route("/")  # Root for login page is index "/"
def login():
    randomiser()  # Create scenario and write all data to txt files
    # Read data from txt files and send it to render
    Info, header = readANS()  # Read data from answer.txt files and send it to render
    manual, mheader = readMan()  # Read data from manual.txt files and send it to render
    # Return render of the login.html page with new local variables
    return render_template(
        "login.html", **locals())


# Control page with route defined by username or username and his order
@app.route("/<username>", methods=['GET', 'POST'])
@app.route("/<username>/<order>/", methods=["GET", "POST"])
def orderClassification(username, order="noorder"):  # By default there is no order


    Info, header = readANS()  # Read data from answer.txt files and send it to render  
    manual, mheader = readMan()  # Read data from manual.txt files and send it to render
    # Taking status of devices, updating it
    statusLamp = check_status(395)
    statusKettle = check_status(19)
    text = ''
    # Greeting answer of Virtual Assistant
    answer = "Hello, " + str(username).title() + ", I am your Virtual Assistant. Type or tell me your order."

    if request.method == 'POST':  # If user POST by clicking submit button any text
        text = request.form['text']
        write2dbLOG("orderraw", text)  # Write to database raw version of order
        try:
            responce, answer, conflict, order = DM.DMorder(text, username)  # Taking data from Dialogue Manager
        except IndexError:  # If system did not managed to classify order
            # Render new answer
            answer = "Sorry, I could not classify your order. Please, try again with different phrase."
            smile = Markup("https://github.com/BiggyBaron/VirtualAssistant/blob/master/static/waiting.gif?raw=true")
            order = "noorder"
            return render_template('main.html', **locals())
        write2dbLOG("orderdef", order)  # Write to database classified version of order
        if responce == 0:  # If DM (dialogue manager) send command 0, then there is a conflict, render yesno.html page
            # Taking status of devices, updating it
            statusLamp = check_status(395)
            statusKettle = check_status(19)
            smile = Markup("https://github.com/BiggyBaron/VirtualAssistant/blob/master/static/yesno.gif?raw=true")
            return render_template('yesno.html', **locals())
        elif responce == 1:  # If DM send command 1, then there is no conflict, device changed, render main.html page
            # Taking status of devices, updating it
            statusLamp = check_status(395)
            statusKettle = check_status(19)
            smile = Markup("https://github.com/BiggyBaron/VirtualAssistant/blob/master/static/sucsses.gif?raw=true")
            return render_template('main.html', **locals())
    # Taking status of devices, updating it
    statusLamp = check_status(395)
    statusKettle = check_status(19)
    smile = Markup("https://github.com/BiggyBaron/VirtualAssistant/blob/master/static/hello.gif?raw=true")
    return render_template('main.html', **locals())


# YesNo page, where user decides go through the conflict or skip it and cancel order
@app.route("/<username>/<order>/yesno", methods=["GET", "POST"])  # Unique route of the page
def yesno(username, order):
    Info, header = readANS()  # Read data from answer.txt files and send it to render
    manual, mheader = readMan()  # Read data from manual.txt files and send it to render
    # Taking status of devices, updating it
    statusLamp = check_status(395)
    statusKettle = check_status(19)
    text = ''
    reason = "noreason"
    if request.method == 'POST':  # If user POST by clicking submit button any text
        text = request.form['text3']
        if text == "no" or text == "No" or text == "Nope":  # If answer is No then render main.html
            answer = "Okay, waiting for orders."
            # Taking status of devices, updating it
            statusLamp = check_status(395)
            statusKettle = check_status(19)
            smile = Markup("https://github.com/BiggyBaron/VirtualAssistant/blob/master/static/waiting.gif?raw=true")
            return render_template('main.html', **locals())
        else:  # Else then render reason.html page and ask about reason
            answer = "Please, define your reason."
            # Taking status of devices, updating it
            statusLamp = check_status(395)
            statusKettle = check_status(19)
            smile = Markup("https://github.com/BiggyBaron/VirtualAssistant/blob/master/static/reason.gif?raw=true")
            return render_template('reason.html', **locals())


# Reason classification page, here system asks user to give a reason for the order
@app.route("/<username>/<reason>/<order>", methods=["GET", "POST"])  # Unique route with data about username and etc.
def reasonClassification(username, order, reason):
    Info, header = readANS()  # Read data from answer.txt files and send it to render
    manual, mheader = readMan()  # Read data from manual.txt files and send it to render
    # Taking status of devices, updating it
    statusLamp = check_status(395)
    statusKettle = check_status(19)
    text = ''
    if request.method == 'POST':  # If user POST by clicking submit button any text
        text = request.form['text2']
        write2dbLOG("reasonraw", text)  # Writing to table raw version of reason
        # Taking status of devices, updating it
        statusLamp = check_status(395)
        statusKettle = check_status(19)
        try:
            answer, reason, smile, responce = DM.DMreason(text, username, order)  # Send to DM and ask render data
        except IndexError:  # If system did not managed to classify reason it asks again
            answer = "Sorry, I could not classify your reason. Please, try again with different phrase."
            smile = Markup("https://github.com/BiggyBaron/VirtualAssistant/blob/master/static/waiting.gif?raw=true")
            reason = "noreason"
            return render_template('reason.html', **locals())
        smile = Markup(smile)
        write2dbLOG("output", responce)  # It writes output of CBR to database
        write2dbLOG("reasondef", reason)  # It writes classified reason to database
        Info, header = Summary()  # Takes summary to show and renders page
    # Taking status of devices, updating it
    statusLamp = check_status(395)
    statusKettle = check_status(19)
    return render_template('summary.html', **locals())


# Main flask app
if __name__ == "__main__":
    # It creates https access by last argument. It is need to be give to web-page permission to microphone
    app.run(host='0.0.0.0', port=8090)  # ssl_context='adhoc')  # If no need in https, just delete last argument
